bin/clone-app

```diff
@@ -0,0 +1,84 @@
+#!/usr/bin/env ruby
+
+class App
+  def initialize(name)
+    @name = name
+  end
+
+  def prepare!
+    Dir.chdir("apps") do
+      if File.directory?(name)
+        update
+      else
+        clone
+      end
+    end
+  end
+
+private
+
+  attr_reader :name
+
+  def clone
+    system "git clone -b #{branch} #{git_url}"
+  end
+
+  def has_local_changes?
+    !system("git diff --quiet --ignore-submodules --no-ext-diff")
+  end
+
+  def checkout
+    system "git checkout #{branch}"
+  end
+
+  def fetch_origin
+    system "git fetch origin"
+  end
+
+  def merge
+    system "git merge --ff-only origin/#{branch}"
+  end
+
+  def update
+    Dir.chdir(name) do
+      if has_local_changes?
+        puts "skipped updating #{name} due to local changes"
+        return true
+      else
+        unless checkout
+          puts "git failed to checkout #{branch}"
+          return false
+        end
+
+        unless fetch_origin
+          puts "git fetch failed for #{name}"
+          return false
+        end
+
+        unless merge
+          puts "update #{name} failed"
+          return false
+        end
+      end
+    end
+    true
```

Comment:
Since these conditions basically terminate the script maybe it'd be cleaner if we had them throw an error in their respective methods? then we could this down to just the method calls rather than conditionals?